Convert to Slash Commands

janine9vn opened the pull request:
It's been awhile and this now looks drastically different! 

# BLUF
This PR completely removes the `!join_ag` !join_wg` commands. Instead, it implements slash command functionality.
Slash Commands provide a better user experience and are easier to parse with the setup. 

There are many changes to allow this to happen but most of them involve replacing functionality 1:1 or removing code that is no longer needed. 

## Docker
This PR also introduces Docker. Lovelace is currently running in an Azure Container Instance. Once this PR is merged and the Dockerfile gets added to the repository, I can begin looking at setting up CI/CD with Azure Containers.

janine9vn committed:
Move constants and restructure directory
- constants moved to constants.py file
- create an overall "bot" directory
- renamed lovelace.py to bot.py

janine9vn committed:
Refactor directory to accomodate extensions
Lovelace is now a package! It requires a different command to run,
the README is updated.
There is now an overaching bot directory folder with appropriate
__init__ and __main__ files.
Additionally, there are two new folder strucutres: exts and utils.
exts is for Extensions while util is for utility functions and classes.
Functionality for autoloading extensions was also added. It checks if
a setup function is in the file, if yes than it'll
load in the extension.

twavv commented on bot/__main__.py:
```diff
@@ -0,0 +1,23 @@
+from bot import exts
+from bot.bot import bot
+
+if __name__ == "__main__":
```
Truthfully, I think this is a bit too "cute" and magical. It would be better™ to just import the packages and then call `.setup()` on them, or have a "extensions" array (and then iterate over them and call `.setup()` on them.

janine9vn replied:
I purposely went with a more "magical" method because I didn't want to have an extensions array we'd have to adjust and change with each extension added. Do you think it'd be better to have call out the different extensions and load them individually? @travigd

twavv replied:
It's prooooooobably fine. Java land does this a bunch with their auto-wired beans.

That being said, I do think it's **better** to just have an array where you specify all of them explicitly.

twavv commented on bot/utils/role_checks.py:
```diff
@@ -0,0 +1,11 @@
+from discord.ext import commands
+
+
+async def _has_role_check(ctx: commands.Context, guild, *role_name: int) -> bool:
```
```suggestion
async def _has_role_check(ctx: commands.Context, guild, *role_names: str) -> bool:
```

It looks like role names are strings, not ints. Also, since this is a varargs, it should be named plurally.

twavv commented on bot/exts/affinity_working_groups.py:
```diff
@@ -1,162 +1,105 @@
+        The current working groups can be found on the oSTEM.org website.
+        """
+        guild = discord.utils.get(bot.guilds, name=ACTIVE_GUILD)
+        if await _has_role_check(ctx, guild, "Member"):
```
Future: consider writing a decorator to check if the user is a member to avoid repeating this auth check in a few different places.

Also, the role name should be a constant.

janine9vn replied:
I'd like to build out the `utils` folder with useful and properly setup decorators, but I was running out of steam to turn the _has_role_check() into a proper decorator. Will add an issue for it to do later.

twavv commented on bot/bot.py:
```diff
@@ -0,0 +1,30 @@
+from bot.constants import COMMAND_PREFIX
+
+
+class oSTEMBot(commands.Bot):
```
Don't have an empty class, just use `command.Bot` directly. Maybe you're intending for this to be customized later, in which case, it's fine.

Nit: class name doesn't conform to CamelCase; consider using `OstemBot` (mixed feelings on this - but I've seen things like `HttpClient` instead of `HTTPClient` so there's convention for just forcing acronyms into CamelCase like that).

janine9vn replied:
I couldn't bring myself to do `OstemBot`, I just couldn't. But the empty class is indeed there for later customization.

twavv reviewed (dismissed):
Nothing blocking, but I left my opinions inline.

I think you **should** address the feedback, but if you merge without doing so, I won't be offended. Especially since the project is still relatively small.

janine9vn committed:
Update bot/utils/role_checks.py

Co-authored-by: Travis DePrato <773453+travigd@users.noreply.github.com>

janine9vn committed:
SImplifying Bot
I've simplified the bot set-up and made it not a module
to make sure that setting it up locally for contributions is easier.

The other big change is the less-smart, but more-readable/understandable
method of loading extensions. If a new extenion is added it must be
manually added to Lovelace's `__init__()`.

These changes are also made in preperation for adding in slash commands.

janine9vn committed:
Add in Docker related files
This adds in the Dockerfile to run a basic instance of this bot.
It takes the bot Token as an ENV argument. Eventually this should be
changed to use a proper secrets file / vault.

janine9vn committed:
Switch functionality to Slash Commands
This probably-too-large commit is a complete overhaul of the
affinity and working group commands to use Slash Commands
instead.

Overall, the previous `!` commands were removed and instead replaced
with the slash commands response.

It makes some assumptions about the server setup and reduces some of the
code complexity and constants required.

With this includes a file to register and update the slash commands.

janine9vn committed:
Remove extraneous file and directory
With the bot simplication, this removes files that are no longer used.

twavv reviewed (approved):
this still shows up as a pending review for me in gh that i want to get rid of,,, so,,, here you go! 😅
